src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1124,17 +974,329 @@
    "ALTER TABLE fact_paths DROP COLUMN value_type_id"
    "ALTER TABLE fact_values DROP COLUMN path_id"))
 
-(defn fill-in-null-producer-timestamp
-  "The producer_timestamp column can sometimes contain null values,
-  especially when updating from older versions. New use cases require
-  a value in that position, so we'll use the value from the
-  'timestamp' column if it's not there. "
+(defn version-2yz-to-300-migration
+  ;; This migration includes:
+  ;;   Insertion of the factsets hash column
+  ;;   Using a report_id (surrogate key) instead of the reports.hash (natural key)
+  ;;   Insert a noop column into reports
+  ;;   Drop latest_report table, merge with certnames
+  ;;   Change name to certname where inconsistent
+  ;;   Insert reports metrics and logs
+  ;;   Ensuring producer_timestamp is NOT NULL
+  ;;   Changing the types of hashes and uuids in postgres to bytea and uuid respectively
```

Comment:
Usually its about reducing the amount of downtime for the upgrade, by simplifying any transformations into as little steps as possible.